crates/noirc_frontend/src/hir/resolution/resolver.rs

```diff
@@ -181,6 +184,7 @@ impl<'a> Resolver<'a> {
         &mut self,
         name: Ident,
         mutable: bool,
+        allow_shadowing: bool,
```

Comment:
It is used when resolving struct patterns. In case of an error resolving the struct type in `let Foo { ... } = ...;` we cannot create a struct pattern as we have no valid struct type to use. So we instead generate an identifier pattern `let $error = ...;` with the name `$error` so that it does not clash with user names. It may still show up in error messages however if there are multiple struct patterns that fail to resolve, and thus we now are trying to have multiple `let $error = ...;` in scope. So we selectively allow shadowing for these cases to hide them from users.

In the future we should also attempt to define the individual field variables as well so that in `let Foo { a, b } = ...;` even if `Foo` fails to resolve, the user won't get errors for `a` and `b` not being defined.